plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/parser/antlr/AbstractAntlrGrammarGenerator.xtend

```diff
@@ -45,30 +45,57 @@ abstract class AbstractAntlrGrammarGenerator {
 	
 	@Inject CodeConfig codeConfig
 	
+	KeywordHelper keyWordHelper
+	
 	def generate(Grammar it, AntlrOptions options, IXtextGeneratorFileSystemAccess fsa) {
+		this.keyWordHelper = KeywordHelper.getHelper(it)
 		val RuleFilter filter = new RuleFilter();
 		filter.discardUnreachableRules = options.skipUnusedRules
-		val RuleNames ruleNames = RuleNames.getRuleNames(grammar, true);
+		val RuleNames ruleNames = RuleNames.getRuleNames(it, true);
 		val Grammar flattened = new FlattenedGrammarAccess(ruleNames, filter).getFlattenedGrammar();
-		fsa.generateFile(grammarNaming.getGrammarClass(it).path + '.g', flattened.compile(options))
+		fsa.generateFile(grammarNaming.getParserGrammar(it).grammarFileName, flattened.compileParser(options))
+		if (!isCombinedGrammar) {
+			fsa.generateFile(grammarNaming.getLexerGrammar(it).grammarFileName, flattened.compileLexer(options))
+		}
+	}
+	
+	protected def isCombinedGrammar() {
+		true
 	}
 	
 	protected abstract def GrammarNaming getGrammarNaming()
 	
-	protected def compile(Grammar it, AntlrOptions options) '''
+	protected def compileParser(Grammar it, AntlrOptions options) '''
 		�codeConfig.fileHeader�
-		grammar �grammarNaming.getGrammarClass(it).simpleName�;
-		�compileOptions(options)�
-		�compileTokens(options)�
-		�compileLexerHeader(options)�
+		�IF !combinedGrammar�parser �ENDIF�grammar �grammarNaming.getParserGrammar(it).simpleName�;
+		�compileParserOptions(options)�
+		�IF isCombinedGrammar�
```

Comment:
Two subsequent checks for `isCombinedGrammar` - is this intentional? Merge both cases?